apps/typegpu-docs/src/content/examples/algorithms/mnist-inference/index.ts

```diff
@@ -0,0 +1,393 @@
+  const buffers = layers.map(([weights, biases]) => {
+    if (weights.shape[1] !== biases.shape[0]) {
+      throw new Error('Shape mismatch');
+    }
+
+    return {
+      weights: weights,
+      biases: biases,
+      buffers: {
+        weights: weights.buffer,
+        biases: biases.buffer,
+        state: root
+          .createBuffer(arrayOf(f32, biases.shape[0]))
+          .$usage('storage'),
+      },
+    };
+  });
+
+  const input = root
+    .createBuffer(arrayOf(f32, layers[0][0].shape[0]))
+    .$usage('storage');
+  const output = buffers[buffers.length - 1].buffers.state;
+
+  const ioBindGroups = buffers.map((_, i) =>
+    root.unwrap(
+      inputOutput.populate({
+        input: i === 0 ? input : buffers[i - 1].buffers.state,
+        output: buffers[i].buffers.state,
+      }),
+    ),
+  );
+
+  const weightsBindGroups = buffers.map((layer) =>
+    root.unwrap(
+      weightsBiases.populate({
+        weights: layer.buffers.weights,
+        biases: layer.buffers.biases,
+      }),
+    ),
+  );
+
+  async function inference(data: number[]): Promise<number[]> {
+    // verify the length of the data matches the input layer
+    if (data.length !== layers[0][0].shape[0]) {
+      throw new Error(
+        `Data length ${data.length} does not match input shape ${layers[0][0].shape[0]}`,
+      );
+    }
+    input.write(data);
+
+    // Run the network
+    const encoder = device.createCommandEncoder();
+    for (let i = 0; i < buffers.length; i++) {
+      const pass = encoder.beginComputePass();
+      pass.setPipeline(pipeline);
+      pass.setBindGroup(0, ioBindGroups[i]);
+      pass.setBindGroup(1, weightsBindGroups[i]);
+      pass.dispatchWorkgroups(buffers[i].biases.shape[0]);
+      pass.end();
+    }
+    device.queue.submit([encoder.finish()]);
+    await device.queue.onSubmittedWorkDone();
+
+    // Read the output
+    return await output.read();
+  }
+
+  return {
+    layers: buffers,
+    input,
+    output,
+    inference,
+  };
+}
+
+// Data fetching and network creation
+
+const [
+  layer0Biases,
+  layer0Weights,
+  layer1Biases,
+  layer1Weights,
+  layer2Biases,
+  layer2Weights,
+] = await Promise.all([
+  fetch('/TypeGPU/mnistWeights/layer0.bias.npy').then((res) =>
+    res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+  ),
+  fetch('/TypeGPU/mnistWeights/layer0.weight.npy').then((res) =>
+    res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+  ),
+  fetch('/TypeGPU/mnistWeights/layer1.bias.npy').then((res) =>
+    res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+  ),
+  fetch('/TypeGPU/mnistWeights/layer1.weight.npy').then((res) =>
+    res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+  ),
+  fetch('/TypeGPU/mnistWeights/layer2.bias.npy').then((res) =>
+    res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+  ),
+  fetch('/TypeGPU/mnistWeights/layer2.weight.npy').then((res) =>
+    res.arrayBuffer().then((buffer) => getLayerData(buffer)),
+  ),
+]);
```

Comment:
`.map` an array of file names (without directories) applying the common fetch/then logic to each element and pass it to `Promise.all` to make this less repetitive